File: problems-extend/3401-3500/3428_maximum-and-minimum-sums-of-at-most-size-k-subsequences/solution.rs
impl Solution {
    pub fn min_max_sums(nums: Vec<i32>, k: i32) -> i32 {
        // After sorting, nums[i] is the maximum of exactly those
        // subsequences whose other members come from the i smaller
        // entries: summed over sizes 1..k that is g(i) = sum_{j < k}
        // C(i, j) subsequences, and symmetrically it is the minimum of
        // g(n-1-i) of them (the larger entries). So the answer is
        // sum nums[i] * (g(i) + g(n-1-i)) mod 10^9 + 7. Each partial row
        // sum rolls in O(1): Pascal gives C(i, j) = C(i-1, j) +
        // C(i-1, j-1), so g(i) = 2*g(i-1) - C(i-1, k-1), one binomial per
        // step from factorial tables. n <= 10^5 keeps those tables small;
        // every residue product stays below ~10^18, inside i64.
        const MOD: i64 = 1_000_000_007;
        let k = k as usize;
        let mut nums = nums;
        nums.sort_unstable();
        let n = nums.len();

        let mut fact = vec![1i64; n];
        for i in 1..n {
            fact[i] = fact[i - 1] * i as i64 % MOD;
        }
        let mut inv_fact = vec![1i64; n];
        inv_fact[n - 1] = Self::mod_pow(fact[n - 1], MOD - 2, MOD);
        for i in (1..n).rev() {
            inv_fact[i - 1] = inv_fact[i] * i as i64 % MOD;
        }
        let choose = |a: usize, b: usize| -> i64 {
            if b > a {
                return 0;
            }
            fact[a] * inv_fact[b] % MOD * inv_fact[a - b] % MOD
        };

        let mut g = vec![0i64; n];
        g[0] = 1;
        for i in 1..n {
            g[i] = (2 * g[i - 1] % MOD + MOD - choose(i - 1, k - 1)) % MOD;
        }
        let mut total = 0i64;
        for i in 0..n {
            total = (total + nums[i] as i64 % MOD * ((g[i] + g[n - 1 - i]) % MOD)) % MOD;
        }
        total as i32
    }

    fn mod_pow(mut base: i64, mut exp: i64, module: i64) -> i64 {
        let mut result = 1i64;
        base %= module;
        while exp > 0 {
            if exp & 1 == 1 {
                result = result * base % module;
            }
            base = base * base % module;
            exp >>= 1;
        }
        result
    }
}

File: problems-extend/3401-3500/3428_maximum-and-minimum-sums-of-at-most-size-k-subsequences/solution.java
import java.util.Arrays;

class Solution {

    // After sorting, nums[i] is the maximum of exactly those subsequences
    // whose other members come from the i smaller entries: summed over
    // sizes 1..k that is g(i) = sum_{j < k} C(i, j) subsequences, and
    // symmetrically it is the minimum of g(n-1-i) of them (the larger
    // entries). So the answer is sum nums[i] * (g(i) + g(n-1-i))
    // mod 10^9 + 7. Each partial row sum rolls in O(1): Pascal gives
    // C(i, j) = C(i-1, j) + C(i-1, j-1), so g(i) = 2*g(i-1) - C(i-1, k-1),
    // one binomial per step from factorial tables. n <= 10^5 keeps those
    // tables small; every residue product stays below ~10^18, inside long.
    public int minMaxSums(int[] nums, int k) {
        long mod = 1_000_000_007L;
        int n = nums.length;
        Arrays.sort(nums);

        long[] fact = new long[n];
        fact[0] = 1;
        for (int i = 1; i < n; i++) {
            fact[i] = (fact[i - 1] * i) % mod;
        }
        long[] invFact = new long[n];
        invFact[n - 1] = modPow(fact[n - 1], mod - 2, mod);
        for (int i = n - 1; i > 0; i--) {
            invFact[i - 1] = (invFact[i] * i) % mod;
        }

        long[] g = new long[n];
        g[0] = 1;
        for (int i = 1; i < n; i++) {
            g[i] = (((2 * g[i - 1]) % mod) + mod - choose(fact, invFact, i - 1, k - 1, mod)) % mod;
        }
        long total = 0;
        for (int i = 0; i < n; i++) {
            total = (total + (nums[i] % mod) * ((g[i] + g[n - 1 - i]) % mod)) % mod;
        }
        return (int) total;
    }

    private static long choose(long[] fact, long[] invFact, int a, int b, long mod) {
        if (b < 0 || b > a) {
            return 0;
        }
        return (((fact[a] * invFact[b]) % mod) * invFact[a - b]) % mod;
    }

    private static long modPow(long base, long exp, long mod) {
        long result = 1;
        long b = base % mod;
        while (exp > 0) {
            if ((exp & 1) == 1) {
                result = (result * b) % mod;
            }
            b = (b * b) % mod;
            exp >>= 1;
        }
        return result;
    }
}
